### PR TITLE
Initial entities implementation

### PR DESCRIPTION
**Pull request by youknowriad**
This PR tries to implement the idea of an entities abstraction. An entity is an abstraction of a WP entity (post, taxonomy, postType, category...). It allows to automatically generate selectors, resolvers for the said entity in the core data module.

In this PR, it's a basic implementation:

 - Entities are static (could be enhanced to be made dynamic)
 - It supports only fetching by PK (could be easily extended to add more selectors)
 - I'm using it for the postType and media entities at the moment.

This PR is blocking #6384 and is related to #6395

### CODE REVIEW

File: core-data/resolvers.js
Comment (jorgefilipecosta) [inline core-data/resolvers.js]: Maybe we should allow model to configure the request as some models probably don't require context=edit.

**Comment by jorgefilipecosta** [comment]
Regarding the functional point of view, these changes test well 👍 
I like this introduction of the model's concept as it could allow us to reduce duplicate code in actions, resolvers, and selectors.
I'm curious about the future plans. Is the idea to involve this model concept? E.g: allow us to mutate and change models besides just retrieving requests? Should we allow the automatic generation of the general request that requests all records e.g: getPostTypes() that requests '/wp/v2/types/', (with opt-in or opt-out flag in the model to enable or disable this). This would allow us to not repeat requests if the "general request" was executed before and now we want a specific record that was already retrieved in the "general request".

**Comment by youknowriad** [comment]
> I'm curious about the future plans. Is the idea to involve this model concept? E.g: allow us to mutate and change models besides just retrieving requests? 

Yes exactly, mutation, pagination, query fetching...

> This would allow us to not repeat requests if the "general request" was executed before and now we want a specific record that was already retrieved in the "general request".

General request => yes, this just sets the baseline but I don't think we should use it to retrieve a specific record because of pagination.

File: core-data/resolvers.js
Comment (youknowriad) [reply core-data/resolvers.js]: I tend to think we should always use the `edit` context (or remove the context concept entirely) but yeah let's keep this way and see if we need to make it configurable as we add models.

File: core-data/reducer.js
Comment (gziolo) [inline core-data/reducer.js]: Can we avoid using `PK` abbreviation and use consistenly `PrimaryKey`? :D

File: core-data/models.js
Comment (gziolo) [inline core-data/models.js]: It should be `type` rather than `kind` for consistency. I guess this is because of the fact that `type` is a reserved word in actions, right?

Comment (youknowriad) [reply core-data/models.js]: Yes it's one of the reasons. Also, type is overused :) so we could have a postype of type type :)

**Comment by gziolo** [comment]
This is cool 👍

File: core-data/models.js
Comment (gziolo) [reply core-data/models.js]: Can we have a post type of kind type? :D

**Comment by youknowriad** [comment]
How can we move forward with this, I'd love to merge it to unblock #6384

**Comment by youknowriad** [comment]
@aduth I'd appreciate your thoughts on this one

**Comment by aduth** [comment]
>and is related to #6395

Can you elaborate on in what way this relates to #6395 ? Do they complement or conflict with one another?

Am planning to take a closer look this evening.

**Comment by aduth** [comment]
What exactly is `kind` ?

File: core-data/reducer.js
Comment (aduth) [inline core-data/reducer.js]: This whole sequence of logic could do for some explaining variables, newlines, comments. Quite difficult to interpret.

File: core-data/selectors.js
Comment (aduth) [inline core-data/selectors.js]: I assume we have some using the existing selector to be updated (work-in-progress) ?

File: core-data/models.js
Comment (aduth) [inline core-data/models.js]: I think you'd at one point floated the idea to me that this configuration could live in state, which would enable others to more easily extend with their own models, and might make it align better with existing concepts like getting via selector, not the standalone utility function here.

File: core-data/models.js
Comment (aduth) [inline core-data/models.js]: I don't know that it's an issue, but I could foresee some confusion on `primaryKey` naming with how it might be confused or misinterpreted as the database concept, where in the case of post types there is no database entity, and for others we might choose a key which isn't actually the primary key for the database entity. Maybe just me?

File: core-data/model-resolvers.js
Comment (aduth) [inline core-data/model-resolvers.js]: Should we organize models-related files into their own folder if there's several related ones which stand apart from the rest of the files?

File: core-data/model-resolvers.js
Comment (aduth) [inline core-data/model-resolvers.js]: Minor: Just noting that this is a runtime evaluation which could theoretically be pre-evaluated (the sort of thing [prepack](https://prepack.io/) would love to optimize for).

File: core-data/reducer.js
Comment (aduth) [inline core-data/reducer.js]: What other properties do you foresee living in this state object?

File: core-data/selectors.js
Comment (jorgefilipecosta) [reply core-data/selectors.js]: This PR now generates dynamically the selector in core-data/model-selectors.js. So everything keeps working as before :)

File: core-data/reducer.js
Comment (youknowriad) [reply core-data/reducer.js]: I'm thinking `queries` maybe, but it could be added later, I can update this PR to remove this key at the moment.

**Comment by youknowriad** [comment]
> What exactly is kind ?

Kind is just a way to group models to avoid duplication, I see these as possible values: `root`, `postType`, `taxonomy`.

This will avoid ambiguity in the reducers and at the selectors level.

> How it relates to #6395

I believe there a conflict in the implementation details but they share similar goes. Instead of having helpers to create reducers/selectors/actions... the models abstraction generate them. #6395 adds the queries data though, and I think this could be refactored to be included in the models abstraction. (each model will generate its `queries` sub state and its selectors).

File: core-data/model-resolvers.js
Comment (youknowriad) [reply core-data/model-resolvers.js]: Move those to the `index.js`. It's not easy to find a good location without creating cycle dependencies.

File: core-data/models.js
Comment (aduth) [reply core-data/models.js]: Maybe just `key` ? Or `keyBy` ?

File: core-data/resolvers.js
Comment (youknowriad) [reply core-data/resolvers.js]: If we use `entityConfig` it means we need an `entity` too, but we don't have such a thing.

I think it's just `entity`. I've used `entityConfig` in the reducer to avoid variable shadowing.

File: core-data/index.js
Comment (gziolo) [inline core-data/index.js]: I would love to see the following method to avoid code duplication and ensure that both selector and resolver names always match:
```js
const getMethodName = ( kind, name ) => {
    const kindPrefix = kind === 'root' ? '' : upperFirst( camelCase( kind ) );
    const nameSuffix = upperFirst( camelCase( name ) );
    return `get${ kindPrefix }${ nameSuffix }`;
};
```
This logic is quite complicated and essential to have it working properly for all selectors so it would be great to have a simple unit tests which ensures it never regresses.

Comment (youknowriad) [reply core-data/index.js]: Good suggestion, updating.

Comment (gziolo) [reply core-data/index.js]: or:

```js
const generateEntityMethods = ( source) => {
	const entityResolvers = entities.reduce( ( memo, { kind, name } ) => {
		const kindPrefix = kind === 'root' ? '' : upperFirst( camelCase( kind ) );
		const nameSuffix = upperFirst( camelCase( name ) );
		return {
			...memo,
			[ `get${ kindPrefix }${ nameSuffix }` ]: ( state, key ) => source.getEntityRecord( state, kind, name, key ),
		};
	}, {} );

const entityResolvesr = generateEntityMethods( resolvers );
const entitySelectors = generateEntityMethods( selectors );
```

**Review by gziolo — APPROVED** [review]
I'm giving 👍 to the implementation. I recommend waiting for another opinion before proceeding given the impact this PR is going to have. I personally love the direction and flexibility this PR introduces. In particular, this will make it very easy to have those entities extensible because of how those definitions are stored in one file: https://github.com/WordPress/gutenberg/blob/9609b27b885598deb495369f28bd19c11c7023a2/core-data/entities.js.

File: core-data/index.js
Comment (aduth) [inline core-data/index.js]: Minor (Perf): As implemented, we're creating a shallow clone on each iteration, when we could be mutating directly.

File: core-data/index.js
Comment (aduth) [inline core-data/index.js]: My DRY senses are tingling 😄 

Maybe:

```js
const createEntityRecordGetter = ( source ) => entities.reduce( ( result, entity ) => {
	const { kind, name } = entity;
	const methodName = getMethodName( kind, name );
	result[ methodName ] = ( state, key ) => resolvers.getEntityRecord( state, kind, name, key );
	return result;
} );
```

File: core-data/reducer.js
Comment (aduth) [inline core-data/reducer.js]: Minor (Perf): We don't need to derive this on every dispatch. It can be in the surrounding function scope.

File: core-data/entities.js
Comment (youknowriad) [reply core-data/entities.js]: I though these shortcuts were getting deprecated from lodash?

File: core-data/entities.js
Comment (youknowriad) [reply core-data/entities.js]: In several methods, the `kind` is the first parameter before the `name` and other optional parameters (think query, for instance). I'd love to get rid of the kind but I fear these methods arguments won't be that easy to handle in that case. `getEntityRecords( null, 'postType', query )`

File: core-data/entities.js
Comment (aduth) [reply core-data/entities.js]: As I understand it, they are on the roadmap to be turned off by default in 5.0.0 .

>Remove shorthand support by default

https://github.com/lodash/lodash/wiki/Roadmap

Doesn't mean they won't exist as an option. I'm personally a fan, and would opt for it.

File: core-data/index.js
Comment (youknowriad) [reply core-data/index.js]: What's your current way of doing this? do you use a specific name, like `resolvers` here for instance?

File: core-data/index.js
Comment (gziolo) [reply core-data/index.js]: I think I saw a similar comment earlier today 😄

Comment (aduth) [reply core-data/index.js]: Oof, the "Show outdated" toggles in GitHub are really easy to miss. 😬

Comment (youknowriad) [reply core-data/index.js]: Since you both asked for it, I executed :)

**Comment by youknowriad** [comment]
Thanks all for your help on this PR

**Pull request merged**
